### ISSUE TITLE
Questions about determinism and input length

### ISSUE THREAD

**Issue by fabiocat93** [tabahi/contexless-phonemes-CUPE#2]
Hi and thanks again for this great repo!

I have a quick question about how contextual information helps the model in its predictions. In particular, I see better results with longer waveforms. Do longer audio inputs help the model? 

If you're open to chat (and maybe explore this together and collaborate!), happy to connect! You can reach me at fabiocat@mit.edu.

Thanks again :)
Fabio

**Comment by tabahi** [comment]
Longer waveforms are generally better for tasks that don’t require sharp contrast between local frames. However, the primary purpose of this model is phoneme alignment for TTS training, which does require sharp contrast for precise boundary detection. 

Extending the waveform length up to 360ms slightly improves phoneme error rate (PER), but at a cost. The model begins to favor longer phonemes over shorter ones, requiring continuous adjustment of the loss function to ensure it learns short phonemes effectively. For example, in the word "gone", /g/ is less than 20ms, while /ɑ/ exceeds 200ms.
In the confusion plot, you'll see that the 120ms context performs better on stop consonants like /t/, /d/, /g/, /p/, and /b/, compared to vowels. This shift is due to increased context contamination in each frame, especially with deeper transformer layers.


This preference for shorter context windows is especially relevant for contextless downstream tasks, such as pure allophones (used in speech style representations) and forced alignment. The idea of "purity" here is analogous to NLP embeddings: a word like "hello" should ideally be represented independently of the next word "world", capturing only the semantics of "hello". But, just as in NLP, some degree of context is necessary to learn better, so it's a trade-off between purity and contextual learning.


I'll email you the paper shortly.

**Comment by fabiocat93** [comment]
Thank you @tabahi , and I am happy to read the paper if you send it over.
 
I truly believe your model has amazing potential even beyond what was initially envisioned: given the efficiency of your model, it could be extremely valuable for real-time timestamped speech-to-viseme detection, especially for syncing audio with mouth movements in lip sync. In the next days, I’ll open-source an agentic AI framework for facilitating the creation of AI agents embodied with avatars. For lip sync, I use your model to process 1s buffer of TTS audio: phonemes map cleanly to visemes at 120ms windows, but this adds ~1s latency. Cutting to 0.5 seconds reduces latency but noticeably degrades phoneme quality. I hear your reasoning about longer and shorter phonemes, and it all makes perfect sense. But do you maybe have any further thoughts on strategies to mitigate the issue I described?

**Comment by tabahi** [comment]
I have tested the latency on Threadrippers and Ampere GPUs. It's less than 200ms for 1 to 10 seconds audio, but it stays the same for no matter the duration on CUDA due to parallel windows. It's better on CPU if you want to convert back to CPU. ~1 second seems too much. Also, the minimum resolution is around 13ms, the context window is 120ms.

I have made it all easier to use in the new aligner repo: [github.com/tabahi/bournemouth-forced-aligner](https://github.com/tabahi/bournemouth-forced-aligner)


Try [this example](https://github.com/tabahi/bournemouth-forced-aligner/blob/main/examples/basic_usage.py), with the [same audio](https://github.com/tabahi/bournemouth-forced-aligner/blob/main/examples/samples/audio/109867__timkahn__butterfly.wav), if it takes more than 0.2s, then something else needs optimization.

**Comment by fabiocat93** [comment]
the computation time is very short indeed (<0.2s), but the performance drops if the input audio is too short (<1s). which is why i was asking your thoughts about working with short audio inputs (<1s), which in a conversational agent also implies a shorter buffer and shorter latency for the user (not latency for the model!!). Thanks again!

**Comment by tabahi** [comment]
Now I understand. Theoretically, chunking 0.5s shouldn't matter since the model only sees 0.12.... It may be an amplitude normalization issue. If 1s has 50% silence, then the RMS nomadization will make the voiced parts more prominent. Longer speech segments tend to have more RMS uniformity,

**Comment by fabiocat93** [comment]
Thanks! I’ll try reducing the buffer to 0.5s, pre-padding 0.5s of silence before inference to stabilize RMS normalization, and finally subtract the 0.5s offset from the predicted timestamps (discarding anything in the padded region)
